Spigot-Server-Patches/0212-Use-TerminalConsoleAppender-for-console-improvements.patch

```diff
@@ -0,0 +1,622 @@
+-        replacements.put(ChatColor.STRIKETHROUGH, Ansi.ansi().a(Attribute.STRIKETHROUGH_ON).toString());
+-        replacements.put(ChatColor.UNDERLINE, Ansi.ansi().a(Attribute.UNDERLINE).toString());
+-        replacements.put(ChatColor.ITALIC, Ansi.ansi().a(Attribute.ITALIC).toString());
+-        replacements.put(ChatColor.RESET, Ansi.ansi().a(Attribute.RESET).toString());
+-    }
+-
+-    @Override
+-    public void sendMessage(String message) {
+-        if (terminal.isAnsiSupported()) {
+-            if (!conversationTracker.isConversingModaly()) {
+-                String result = message;
+-                for (ChatColor color : colors) {
+-                    if (replacements.containsKey(color)) {
+-                        result = result.replaceAll("(?i)" + color.toString(), replacements.get(color));
+-                    } else {
+-                        result = result.replaceAll("(?i)" + color.toString(), "");
+-                    }
+-                }
+-                System.out.println(result + Ansi.ansi().reset().toString());
+-            }
+-        } else {
+-            super.sendMessage(message);
+-        }
+-    }
+-
+-    public static ConsoleCommandSender getInstance() {
+-        if (Bukkit.getConsoleSender() != null) {
+-            return Bukkit.getConsoleSender();
+-        } else {
+-            return new ColouredConsoleSender();
+-        }
+-    }
+-}
+diff --git a/src/main/java/org/bukkit/craftbukkit/command/ConsoleCommandCompleter.java b/src/main/java/org/bukkit/craftbukkit/command/ConsoleCommandCompleter.java
+index 33e8ea02..7979cd1a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/command/ConsoleCommandCompleter.java
++++ b/src/main/java/org/bukkit/craftbukkit/command/ConsoleCommandCompleter.java
+@@ -8,17 +8,27 @@ import java.util.logging.Level;
+ import org.bukkit.craftbukkit.CraftServer;
+ import org.bukkit.craftbukkit.util.Waitable;
+ 
+-import jline.console.completer.Completer;
++// Paper start - JLine update
++import net.minecraft.server.DedicatedServer; // Paper
++import org.jline.reader.Candidate;
++import org.jline.reader.Completer;
++import org.jline.reader.LineReader;
++import org.jline.reader.ParsedLine;
++// Paper end
+ import org.bukkit.event.server.TabCompleteEvent;
+ 
+ public class ConsoleCommandCompleter implements Completer {
+-    private final CraftServer server;
++    private final DedicatedServer server; // Paper - CraftServer -> DedicatedServer
+ 
+-    public ConsoleCommandCompleter(CraftServer server) {
++    public ConsoleCommandCompleter(DedicatedServer server) { // Paper - CraftServer -> DedicatedServer
+         this.server = server;
+     }
+ 
+-    public int complete(final String buffer, final int cursor, final List<CharSequence> candidates) {
++    // Paper start - Change method signature for JLine update
++    public void complete(LineReader reader, ParsedLine line, List<Candidate> candidates) {
++        final CraftServer server = this.server.server;
```

Comment:
@Zbob750 

> Was this added just so you can stop using `this` everywhere?

No, the command completer is now initialized earlier, directly from the console thread in `DedicatedServer`. At this point, `CraftServer` is not available yet so I had to change the field in this class from `CraftServer` to `DedicatedServer`. The purpose of this variable is to keep the existing code as original as possible because it references the `CraftServer` field. It works for some usages (e.g. the ones in the `Waitable`) but not for the usages prefixed with `this.`.